### PR TITLE
add a specific "documentation_formats" that may differ from the API "…

### LINKED ISSUE #1854: Add a `documentation_formats` that differs from `formats`
If you de-activate the `json` format from the configuration, you can not use like [ReDoc](https://rebilly.github.io/ReDoc/). You can not use the integrated swagger if you disable `html` format.

Personally, I disable everything except `json-ld` (less options means less bugs ;) ).

Furthermore, I don't think that enabling html format is really pertinent for an API, but it does for the documentation.

Do you think it's a good idea to split `formats` from the `documentation_formats`, this way, we could do something like:

```yaml
api_platform:
    formats: 
        jsonld: ['application/ld+json']
    documentation_formats: 
        html: ['text/html']
        json: ['application/json']
```

As usual, I will open a PR if you agree with this.

### PR DESCRIPTION
**Pull request by jdeniau**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | See descrption
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1854 
| License       | MIT
| Doc PR        | https://github.com/api-platform/docs/pull/475

This PR allow to disable the "json" format from the api and only allow jsonld for example, without breaking the swagger documentation.

About the deprecation though: I removed "json" and "html" from the formats, to keep only "jsonld", and moved "json" and "html" to the documentation formats only.
Technically, it's a compatibility break, but as often can be seen as a bugfix (at least for the `html` format). 

If you like, I can add the `json` format back, but as I see it, api-platform default is really nicely coupled with json-jd, so I assumed that the `json` format it was for documentation only.

### CODE REVIEW

File: src/EventListener/AddFormatListener.php [hunk 1]
```diff
@@ -28,12 +28,14 @@
 {
     private $negotiator;
     private $formats;
+    private $documentationFormats;
     private $mimeTypes;
 
-    public function __construct(Negotiator $negotiator, array $formats)
+    public function __construct(Negotiator $negotiator, array $formats, array $documentationFormats)
```
Comment (dunglas) [inline src/EventListener/AddFormatListener.php]: It must have a default value (`= []`), or it's BC break.

File: src/EventListener/AddFormatListener.php [hunk 2]
```diff
@@ -145,4 +151,17 @@ private function getNotAcceptableHttpException(string $accept, array $mimeTypes
             implode('", "', $mimeTypes)
         ));
     }
+
+    private function getRequestFormats(Request $request): array
+    {
+        return $this->isDocumentation($request)
```
Comment (dunglas) [inline src/EventListener/AddFormatListener.php]: Can you inline this please?

File: src/EventListener/AddFormatListener.php [hunk 3]
```diff
@@ -145,4 +151,17 @@ private function getNotAcceptableHttpException(string $accept, array $mimeTypes
...
+            ? $this->documentationFormats
+            : $this->formats
+        ;
+    }
+
+    private function isDocumentation(Request $request): bool
```
Comment (dunglas) [inline src/EventListener/AddFormatListener.php]: Can you remove this function? (It's in the hot path, and function calls are costly).

File: src/EventListener/AddFormatListener.php [hunk 4]
```diff
@@ -145,4 +151,17 @@ private function getNotAcceptableHttpException(string $accept, array $mimeTypes
...
+    {
+        return 'api_platform.action.documentation' === $request->attributes->get('_controller');
```
Comment (dunglas) [inline src/EventListener/AddFormatListener.php]: Alternative proposal:

set a `_api_documentation` attribute here https://github.com/api-platform/core/blob/master/src/Bridge/Symfony/Bundle/Resources/config/routing/docs.xml#L10, and test for the presence of this flag. It would be more flexible and give an extension point.

Comment (jdeniau) [reply src/EventListener/AddFormatListener.php]: @dunglas What you think about a `_api_endpoint_type` set to `documentation` instead ? It may avoid anyone to add another variable later.

Comment (dunglas) [reply src/EventListener/AddFormatListener.php]: I usually prefer boolean flags (almost all others are boolean). What other type can it be?

File: src/Bridge/Symfony/Bundle/EventListener/SwaggerUiListener.php [hunk 5]
```diff
@@ -34,4 +34,10 @@ public function onKernelRequest(GetResponseEvent $event)
 
         $request->attributes->set('_controller', 'api_platform.swagger.action.ui');
     }
+
+    public function addEndpointTypeToRequest(GetResponseEvent $event)
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/EventListener/SwaggerUiListener.php]: Is it necessary? The flag will always be set for every request even if it's not caught by Swagger UI isn't it?

**Comment by jdeniau** [comment]
@dunglas done, I just changed `_api_documentation` by `_api_endpoint_type` set to `documentation`.

For the circleci test, I think it's a problem with circleci configuration

File: src/Bridge/Symfony/Bundle/EventListener/SwaggerUiListener.php (hunk 5 above)
Comment (jdeniau) [reply src/Bridge/Symfony/Bundle/EventListener/SwaggerUiListener.php]: Oh, I need to set the same check as `onKernelRequest` function.

I need to separate the two listener because the AddFormatListener come between priority 0 and 7

File: src/EventListener/AddFormatListener.php (hunk 4 above)
Comment (jdeniau) [reply src/EventListener/AddFormatListener.php]: may something related to graphql :man_shrugging:

**Comment by jdeniau** [comment]
@dunglas I'm kind of stuck here with a circular exception on the SwaggerUiListener, maybe you have an idea :

The SwaggerUiListener depends on the `AddFormatListener` for the `text/html` request format to be set.

The AddFormatListener need to know if we are on a documentation page or not, and thus needs to have the documentation attribute set.

The only solution I see here is to dissociate the AddFormatListener and something that throw the exception but it seems really overkill for such a small feature, don't yout think ?

**Comment by dunglas** [comment]
I would handle Swagger UI in two different ways:

* For the `/docs` URL: handle it using `documentation_formats` as done in this PR
* For all others URL, consider that HTML is **not** a "documentation" format (because it is not really, it allows to serialize the resource itself as HTML). So there are no need to mark all those routes as "documentation" (they are not, they are resources routes). I would just let the hijacking mechanism working as it is currently, except that I suggest to disable the hijacking if Swagger is not present in `documentation_formats`. To clarify: disable the HTML format if Swagger is not enabled in the DI extension.

WDYT?

**Comment by dunglas** [comment]
Any news on this one?

**Comment by jdeniau** [comment]
Sorry, I was on holidays ;)

I will check your return soon.

**Comment by dunglas** [comment]
Any news :)

**Comment by jdeniau** [comment]
@dunglas should be OK.

I removed the weird listener that I previously added.

As it is very internal functionality, I am not 100% sure of what I have done though ;)

File: src/Bridge/Symfony/Bundle/Resources/config/swagger-ui.xml [hunk 6]
```diff
@@ -7,7 +7,7 @@
     <services>
 
         <service id="api_platform.swagger.listener.ui" class="ApiPlatform\Core\Bridge\Symfony\Bundle\EventListener\SwaggerUiListener">
-            <tag name="kernel.event_listener" event="kernel.request" method="onKernelRequest" />
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/swagger-ui.xml]: Why this change? 0 is the default priority.

**Comment by jdeniau** [comment]
Still an issue due to changes I made in the available formats

**Comment by jdeniau** [comment]
@dunglas everything is good

**Comment by jdeniau** [comment]
My turn to ping you @dunglas ;)

Fixes #1854

File: tests/EventListener/AddFormatListenerTest.php [hunk 7]
```diff
@@ -32,7 +32,7 @@ public function testNoResourceClass()
         $eventProphecy->getRequest()->willReturn($request)->shouldBeCalled();
         $event = $eventProphecy->reveal();
 
-        $listener = new AddFormatListener(new Negotiator(), ['notexist' => 'application/vnd.notexist']);
+        $listener = new AddFormatListener(new Negotiator(), ['notexist' => 'application/vnd.notexist'], []);
```
Comment (soyuka) [inline tests/EventListener/AddFormatListenerTest.php]: As the new array argument has a default value we don't really need to change this right?

File: features/main/content_negotiation.feature [hunk 8]
```diff
@@ -110,8 +110,8 @@ Feature: Content Negotiation support
     Then the response status code should be 406
     And the header "Content-Type" should be equal to "application/problem+json; charset=utf-8"
 
-  Scenario: If the request format is HTML, the error should be in HTML
```
Comment (dunglas) [inline features/main/content_negotiation.feature]: Can you keep this text (and duplicate it for XML if appropriate)? It checks that the documentation for any URL feature works (for instance when you open `http://localhost/books/1`, Swagger UI shows up and automatically runs the AJAX request).

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -222,6 +222,9 @@ public function getConfigTreeBuilder()
 
         $this->addFormatSection($rootNode, 'formats', [
             'jsonld' => ['mime_types' => ['application/ld+json']],
+        ]);
+        $this->addFormatSection($rootNode, 'documentation_formats', [
+            'jsonld' => ['mime_types' => ['application/ld+json']], // Hydra documentation, admin and PWA generator
             'json' => ['mime_types' => ['application/json']], // Swagger support
             'html' => ['mime_types' => ['text/html']], // Swagger UI support
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: HTML must be defined also in `addFormatSection` section to have the "documentation for any URL" feature working.

File: src/EventListener/AddFormatListener.php [hunk 10]
```diff
@@ -145,4 +151,16 @@ private function getNotAcceptableHttpException(string $accept, array $mimeTypes
             implode('", "', $mimeTypes)
         ));
     }
+
+    /**
+     * Retrieves the list of accepted format in the request depending on some parameters.
+     */
+    private function getRequestAcceptedFormats(Request $request): array
+    {
+        if (!$request->attributes->has('_api_endpoint_type')) {
```
Comment (dunglas) [inline src/EventListener/AddFormatListener.php]: you can call `get` directly and store the result in a variable to prevent an extra function call. (get will return `null` if the attribute is not defined)

File: src/EventListener/AddFormatListener.php [hunk 11]
```diff
@@ -51,13 +53,15 @@ public function onKernelRequest(GetResponseEvent $event)
             return;
         }
 
-        $this->populateMimeTypes();
-        $this->addRequestFormats($request, $this->formats);
+        $requestAcceptedFormats = $this->getRequestAcceptedFormats($request);
```
Comment (dunglas) [inline src/EventListener/AddFormatListener.php]: Actually, this private function can be replaced by `'documentation' === $request->attributes->get('_api_endpoint_type') ? $this->documentationFormats : $this->formats`. It removes 2 function calls in the hot path.

File: src/EventListener/AddFormatListener.php [hunk 12]
```diff
@@ -84,11 +88,11 @@ public function onKernelRequest(GetResponseEvent $event)
                 return;
             }
 
-            throw $this->getNotAcceptableHttpException($mimeType);
+            throw $this->getNotAcceptableHttpException($mimeType ?? $requestFormat);
```
Comment (dunglas) [inline src/EventListener/AddFormatListener.php]: `$requestAcceptedFormats` right? Thus should be covered by a test (because the error isn't caught now)

**Comment by jdeniau** [comment]
I'm kinda lost here because of #1798 : it added a format resolver but does not include error format. 

Does documentation format should be in `FormatsProvider` ? What do you thing @antograssiot ?

**Comment by antograssiot** [comment]
Indeed there is no ErrorFormatsProvider (yet ?).
I can add it if there is a need, it should be pretty easy, and I guess that if you feel that document format needs in some cases to be specified at resource/operation level, you could use a DocumentFormatsProvider or something similar but it seems less likely to be used IMO.

Let's see maintainers opinion.

**Comment by dunglas** [comment]
@jdeniau, do you need any help to finish this PR?

**Comment by jdeniau** [comment]
Actually I think I do yes!
I am kind of lost here, because I don't know api-platform internals as you guys do. 
Maybe I should look back at #1798 and try harder but any help would be appreciated!

**Comment by flug** [comment]
@jdeniau Up is this pull request still valid?

**Comment by jdeniau** [comment]
I did not manage to fix this and have no motivation on that now. I am closing this one.
